### PR TITLE
Added covinient methods to manage node modules as adapters

### LINKED ISSUE #2006: RFC: Install user-defined dependencies centrally, instead of letting adapters handle them
"Manually" installing dependencies in a sub-package of our "install" project (`/opt/iobroker/package.json`) once worked fine, but increasingly collides with what `npm` does while managing the dependencies. We already stopped doing a separate `npm install` inside the adapter directories in https://github.com/ioBroker/ioBroker.js-controller/pull/1339, but some adapters still do this, leading to [weird issues](https://github.com/ioBroker/ioBroker.javascript/issues/1146#issuecomment-1251959013).

I therefore propose we get rid of this too and do the installation in a central location, adding the dependencies to our root `package.json`. To avoid conflicts, we need to fake-namespace these packages, which npm supports using the `npm:` protocol. Npm supports this since `6.9.0` (https://github.com/npm/cli/pull/3):
```jsonc
// package.json
"dependencies": {
  //... all adapters etc., as usual
  // namespaced user-defined dependencies of iobroker.javascript, instance 0:
  "@iobroker-javascript.0/axios": "npm:axios@1.2.3", // Here javascript.0 has axios@1.2.3 as its dependency
  "@iobroker-javascript.0/alcalzone-pak": "npm:@alcalzone/pak@0.9.0", // scoped dependencies need to be escaped. I propose omitting the leading @ and replacing / with - (_ would also be fine)
  // ... likewise for other adapters/instances
}
```

This has the downside that the adapters need to be able to redirect the `require` to the correct package name, e.g. `javascript.0` would have to require `@iobroker-javascript.0/axios` instead of just `axios`, but we can expose a utility for this too.

The upside would be that versioning these dependencies would become a bit more manageable, and another adapter installing the same one would not accidentally cause a conflict. This way we can also support multiple instances of `javascript` have different versions of the same package, since the fake scopes are instance-specific. In addition, maybe the adapters wouldn't even have to store their dependencies separately anymore, but could rather enumerate them from the root `package.json`.

### PR DESCRIPTION
**Pull request by foxriver76**
closes #2006

### CODE REVIEW

File: packages/cli/src/lib/setup.ts [hunk 1]
```diff
@@ -853,28 +854,20 @@
...
             }
             if (parseInt(repoUrl, 10).toString() === (repoUrl || '').toString()) {
                 const temp = instance;
                 instance = repoUrl;
                 repoUrl = temp;
             }
-            if (parseInt(instance, 10).toString() === (instance || '').toString()) {
-                instance = parseInt(instance, 10);
-                params.instance = instance;
+            if (instance && parseInt(instance, 10).toString() === (instance || '').toString()) {
```
[the remark this answers is not in the document]
Comment (foxriver76) [reply packages/cli/src/lib/setup.ts:869]: Instance can be `string` | `undefined` at this place according to ts, thus the guard is needed for successful build

**Review by Apollon77 — APPROVED** [review]
Readmewould be missing and then maybe a javascript adapter ticket to somehow validate this feature

**Review by AlCalzone — COMMENTED** [review]
Do we also need a way to enumerate packages installed for an adapter or let an adapter check whether a given package is installed?

And we might also need a way to resolve the actual path of the module, at least for the TypeScript support in the JS adapter.

**Comment by seb2010** [comment]
Morning all,
I am highly appreciating managing the packages now seperately per instance. Hope that solves my frequent problems after adpater upgrades.
One thing to add: the log output of the npm messages is now not available anymore. So any problems on this side will not be visible to the admin anymore. Is it possible to output the package related npm output in the logs again?
KR SEB

**Comment by foxriver76** [comment]
The install method returns stdout and stderr, Adapters are able to log this if desired, so would need to be done eg by JavaScript which uses the method.

**Pull request merged**